File: backends/kotlin-ktor/src/main/kotlin/com/authorization/handler/GateHandler.kt
/*
 * ゲート HTTP ハンドラーモジュール。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
package com.authorization.handler

import com.authorization.usecase.gate.Interactor as GateUC
import com.authorization.usecase.gate.IssueDto
import com.authorization.usecase.gate.VerifyDto
import io.ktor.http.*
import io.ktor.server.application.*
import io.ktor.server.response.*
import kotlinx.serialization.json.*

/**
 * ゲート認可 API のハンドラーです。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
class GateHandler(private val gateUC: GateUC) {

    /**
     * JWT トークンを発行します。
     *
     * @param call アプリケーションコール
     */
    suspend fun issue(call: ApplicationCall) {
        val member = call.request.queryParameters["member"]
        if (member.isNullOrBlank()) {
            call.respond(HttpStatusCode.BadRequest, buildJsonObject { put("error", "member_required") })
            return
        }
        val auth = call.request.headers["Authorization"] ?: ""
        val accessToken = if (auth.startsWith("Bearer ")) auth.removePrefix("Bearer ") else ""
        val vo = gateUC.issueToken(IssueDto(accessToken = accessToken, memberId = member))
        call.respond(buildJsonObject { put("token", vo.token) })
    }

    /**
     * JWT トークンを検証します。
     *
     * @param call アプリケーションコール
     */
    suspend fun verify(call: ApplicationCall) {
        val identifier = call.parameters["identifier"] ?: ""
        val token = call.request.queryParameters["token"]
        if (token.isNullOrBlank()) {
            call.respond(HttpStatusCode.BadRequest, buildJsonObject { put("error", "token_required") })
            return
        }
        val vo = gateUC.verify(VerifyDto(identifier = identifier, token = token))
        call.respond(buildJsonObject {
            vo.claims.forEach { (k, v) ->
                when (v) {
                    null       -> put(k, JsonNull)
                    is Boolean -> put(k, v)
                    is Long    -> put(k, v)
                    is Double  -> put(k, v)
                    else       -> put(k, v.toString())
                }
            }
        })
    }
}

File: backends/go-echo/internal/handler/gate_handler.go
package handler

import (
	ugate "authorization-go-echo/internal/usecase/gate"
	"authorization-go-echo/pkg/apperror"
	"net/http"
	"strings"

	"github.com/labstack/echo/v4"
)

type GateHandler struct {
	gateUC *ugate.Interactor
}

func NewGateHandler(gateUC *ugate.Interactor) *GateHandler {
	return &GateHandler{gateUC: gateUC}
}

func (h *GateHandler) Issue(c echo.Context) error {
	member := c.QueryParam("member")
	if member == "" {
		return apperror.BadRequest("member_required")
	}

	auth := c.Request().Header.Get("Authorization")
	accessToken := strings.TrimPrefix(auth, "Bearer ")

	vo, err := h.gateUC.IssueToken(ugate.IssueDto{
		AccessToken: accessToken,
		MemberID:    member,
	})
	if err != nil {
		return err
	}
	return c.JSON(http.StatusOK, map[string]interface{}{"token": vo.Token})
}

func (h *GateHandler) Verify(c echo.Context) error {
	identifier := c.Param("identifier")
	token := c.QueryParam("token")
	if token == "" {
		return apperror.BadRequest("token_required")
	}

	vo, err := h.gateUC.Verify(ugate.VerifyDto{
		Identifier: identifier,
		Token:      token,
	})
	if err != nil {
		return err
	}
	return c.JSON(http.StatusOK, vo.Claims)
}
